xarray/core/computation.py

```diff
@@ -1326,12 +1326,23 @@ def _cov_corr(da_a, da_b, dim=None, ddof=0, method=None):
 
     # 2. Ignore the nans
     valid_values = da_a.notnull() & da_b.notnull()
+    valid_count = valid_values.sum(dim) - ddof
 
-    if not valid_values.all():
-        da_a = da_a.where(valid_values)
-        da_b = da_b.where(valid_values)
+    def _get_valid_values(da, other):
+        """
+        Function to lazily mask da_a and da_b
+        following a similar approach to
+        https://github.com/pydata/xarray/pull/4559
+        """
+        missing_vals = np.logical_or(da.isnull(), other.isnull())
+        if missing_vals.any():
+            da = da.where(~missing_vals)
+            return da
+        else:
+            return da
 
-    valid_count = valid_values.sum(dim) - ddof
+    da_a = da_a.map_blocks(_get_valid_values, args=[da_b])
```

Comment:
I think it is possible to only run this function once.
``` python
da_a, da_b = xr.apply_ufunc(_get_valid_values, da_a, da_b, ..., dask="parallelized")
```
and replace `.isnull` with `duck_array_ops.isnull` in `_get_valid_values`.

We can also do this in a future PR if you'd like. 

Right now, this PR just needs some tests to make sure every thing is lazy and that the check works.

Reply:
Cool! I'll try writing a test soon. Would it be ok to follow the tests that were implemented [here](https://github.com/pydata/xarray/pull/4559/files#diff-74d2dc289aa601b2de094fb3a3b687fd65963401b51b95cc5e0afcd06cc4cb82)? With `raise_if_dask_computes`?

Reply:
sure, sounds good